Rework the auto low luck casualty selection

trevan opened the pull request:
Rewrite the old comments to fit better with how the game now works. When
this was first worked on, all AA were infinite, the dice sides was
always 6, and AA only had a strength of 1 or 2.

The new logic is a little different from the old logic. The old logic
used a little more randomness than the new logic.  But that randomness
doesn't appear to improve the functionality so it was removed.


<!--
  Commit comment above summarizing the update.  If multiple commits please
  summarize the change above.
  Code standards and PR guidelines can be found at:
  <https://github.com/triplea-game/triplea/wiki/Contribution-Guidelines>
-->

## Testing
<!-- Describe any manual testing performed below. -->
Ran Hard AI of Warcraft and TWW

## Screens Shots
<!-- If there are UI updates, include screenshots below -->

## Additional Notes to Reviewer
<!-- Add any additional details that would be helpful to reviewers -->

## Release Note

<!--
Include a release note if there is a bug fix or a visible change for players.
For format & syntax help, see:
https://github.com/triplea-game/triplea/blob/master/docs/pr-release-notes.md
-->

<!--RELEASE_NOTE--><!--END_RELEASE_NOTE-->

trevan commented:
The unit test failures don't appear to be related to this PR.

DanVanAtta commented on game-core/src/main/java/games/strategy/triplea/delegate/battle/casualty/AaCasualtySelector.java:
```diff
@@ -114,265 +112,209 @@ private static CasualtyDetails getLowLuckAaCasualties(
-          hitTargets.add(i);
-        }
+    final int guaranteeHitGroupSize =
+        calculateGuaranteeLowLuckHitGroupSize(targets, diceRoll, unitPowerAndRollsMap);
```
Is this a place where we could early terminate and avoid processing on empty lists?

EG:
```
    final int guaranteeHitGroupSize =
        calculateGuaranteeLowLuckHitGroupSize(targets, diceRoll, unitPowerAndRollsMap);
   if(guaranteedHitGroupSize == 0) {
           return LowLuckTargetGroups.NO_GUARANTEED_HITS;
    }
```

Note, this comment was made out of order for other comments below, this could moot some of those suggestions.

DanVanAtta reviewed (commented):
I made it about half way through this update. I hope to return to this perhaps tomorrow or Sunday evening to finish up.

Overall my impression is quite positive, things are for sure becoming simpler & more clear. You're certainly brave to be diving into this code @trevan .

DanVanAtta commented on game-core/src/main/java/games/strategy/triplea/delegate/battle/casualty/AaCasualtySelector.java:
```diff
@@ -114,265 +112,209 @@ private static CasualtyDetails getLowLuckAaCasualties(
+    final int bestStrength = unitPowerAndRollsMap.getBestStrength();
+    final int chosenDiceSize = unitPowerAndRollsMap.getDiceSides();
+
+    final boolean hasOverstackHits =
```
It's not immediately clear what an overstack hit is. Maybe a quick comment explaining the term, or a different variable name would make this clear? I struggled here a bit and then reading a bit more later about the overstack AA hits it made more sense (but I'm still not sure I understand the concept TBH).

DanVanAtta commented on game-core/src/main/java/games/strategy/triplea/delegate/battle/casualty/AaCasualtySelector.java:
```diff
@@ -114,265 +112,209 @@ private static CasualtyDetails getLowLuckAaCasualties(
     }
-    return Tuple.of(groupsOfSize, toRoll);
+
+    return buildCasualtyDetails(availableTargets, hitTargets);
```
Perhaps a style preference on my part, to another extent I see we are doing the 'single-return' point design as well. I think some this method block would be relatively clear returning when we can. For example, the last 'else' statement would avoid some extra thought to realize that `availableTargets' is equal to 'hitTargets'.

Personally I've come to find the single-return point design is only really good for longer methods, ones that are complicated and return for different reasons at different places. For short methods, something like the below is just fine IMO:
```
if
   return
else if
   return
else
   return
```

The above example I'd say is more clear than the single return variable, ie:
```
var returnVal;
if
   returnVal = a
else if
   returnVal =b
else
   returnVal =c

return returnVal
```

DanVanAtta commented on game-core/src/main/java/games/strategy/triplea/delegate/battle/casualty/AaCasualtySelector.java:
```diff
@@ -114,265 +112,209 @@ private static CasualtyDetails getLowLuckAaCasualties(
+  /** Find the targets that were hit by a dice roll */
+  private static Collection<Unit> findRolledTargets(
+      final List<Unit> availableTargets, final DiceRoll dice) {
+    final List<Die> rolls = dice.getRolls();
```
Would it be too nitpicky to question why there is an intermediate `rolls` variable? `dice.getRolls()` is just as descriptive and it's only used a couple times below. Removing intermediate variables is often a good thing, one less thing for a maintainer to keep on their mental stack when reading code. At this point when reading the method, there are 3 variables in scope, soon to be a 4th one, leaves little room for much else!

DanVanAtta commented on game-core/src/main/java/games/strategy/triplea/delegate/battle/casualty/AaCasualtySelector.java:
```diff
@@ -114,265 +112,209 @@ private static CasualtyDetails getLowLuckAaCasualties(
+    final List<Unit> hitTargets = new ArrayList<>();
+    for (int i = 0; i < rolls.size(); i++) {
+      if (rolls.get(i).getType() == DieType.HIT) {
+        hitTargets.add(availableTargets.get(i));
```
It looks like there is an implicit assumption that avaialbleTargets will have the same size as 'dice.getRolls'. This assumption is not necessarily clear here, a `Precondition.checkArgument(availableTarget.size() == dice.getRolls().size())` at the start of the method would help document this. A reader would do a double take here wondering if there could be an indexing problem because `i` is tracking against `rolls`, not `availableTargets`.

DanVanAtta reviewed (commented):
Finished looking over this. The updates does make for better code!
I left a number of design suggestions and a few nits, nothing too critical or that I would insist on, all suggestions. Please have a look @trevan and incorporate what makes sense and feel free to discuss further please.

trevan commented:
@DanVanAtta , I like the suggestion to make the `LowLuckTargetGroups` a more complete object.  I've moved it to its own file, moved the "construction" code to it and made it a true constructor and encapsulated a little more logic in it.

I also changed where the `buildCasualtyDetails` is called so that the methods can return earlier.
